src/main.rs

```diff
@@ -0,0 +1,380 @@
+                }
+            })
+        }
+
+        let paths_to_patterns = |paths: &Vec<String>| {
+            make_absolute_paths(&cwd, paths)
+                .filter_map(|path| {
+                    Pattern::new(path.to_string_lossy().as_ref()).map_or_else(
+                        |e| {
+                            eprintln!("invalid glob pattern: {e:?}");
+                            None
+                        },
+                        |pat| Some(pat),
+                    )
+                })
+                .collect()
+        };
+
+        let prefixes: Vec<_> = make_absolute_paths(&cwd, &req.patterns).collect();
+        let patterns = paths_to_patterns(&req.patterns);
+        let ignores = paths_to_patterns(&req.ignores);
+
+        let events = req.events;
+
+        WatcherConfig {
+            cwd,
+            events,
+            ignores,
+            patterns,
+            prefixes,
+        }
+    }
+}
+
+fn normalize_events(events: &mut Vec<notify::Event>) {
+    use notify::event::{CreateKind, EventAttributes, ModifyKind, RemoveKind, RenameMode};
+    use notify::{Event, EventKind};
+
+    let mut i = 0;
+    while i < events.len() {
+        let event = &mut events[i];
+        if let EventKind::Modify(ModifyKind::Name(rename)) = event.kind {
+            match rename {
+                RenameMode::From => {
+                    event.kind = EventKind::Remove(RemoveKind::Any);
+                }
+                RenameMode::To => {
+                    event.kind = EventKind::Create(CreateKind::Any);
+                }
+                RenameMode::Both => {
+                    assert_eq!(event.paths.len(), 2);
+                    event.kind = EventKind::Remove(RemoveKind::Any);
+                    let dest = event.paths.pop().unwrap();
+                    events.insert(
+                        i + 1,
+                        Event {
+                            kind: EventKind::Modify(ModifyKind::Name(RenameMode::To)),
+                            paths: vec![dest],
+                            attrs: EventAttributes::new(),
+                        },
+                    )
+                }
+                _ => (),
+            }
+        }
+        i += 1;
+    }
+}
+
+fn event_handler(configs: &Mutex<BTreeMap<usize, WatcherConfig>>, events: DebounceEventResult) {
+    let mut events = match events {
+        Ok(events) => events.into_iter().map(|event| event.event).collect(),
+        Err(errors) => {
+            for e in errors {
+                eprintln!("watcher error: {e:?}");
+            }
+            return;
+        }
+    };
+
+    normalize_events(&mut events);
+
+    let mut stdout = BufWriter::new(io::stdout().lock());
+    let mut written = false;
+    for event in events {
+        let event_type = match event.kind {
+            notify::EventKind::Create(_) => EventType::Create,
+            notify::EventKind::Modify(_) => EventType::Change,
```

Comment:
The reason for double notification on editing a file is the fact that we are being notified twice with two different event attributes - once with `Modify(Metadata(Any))` and once with `Modify(Data(Content))`:

```
event: Event { kind: Modify(Metadata(Any)), paths: ["/usr/local/workspace/sublime-packages/LSP-file-watcher-rust/src/main.rs"], attr:tracker: None, attr:flag: None, attr:info: None, attr:source: None }
event: Event { kind: Modify(Data(Content)), paths: ["/usr/local/workspace/sublime-packages/LSP-file-watcher-rust/src/main.rs"], attr:tracker: None, attr:flag: None, attr:info: None, attr:source: None }
```

I believe that the correct thing to do would be to filter out any non-Content ones since LSP only wants to get content notification changes.

```suggestion
            notify::EventKind::Modify(ModifyKind::Data(DataChange::Content)) => EventType::Change,
```

(also requires adding some `use` statements)

Reply:
I tested chokidar and found it reported metadata changes made with e.g. `touch` as well. So the problem was not not ignoring metadata changes but was not merging events. Fixed in 5eb580c. Please have a look.